File: bridges/nvda/addon/globalPlugins/nvdaMcpBridge/domain/controllers/commands/get_speech.py
# nvdaMcpBridge domain -- GetSpeechHandler: speech captured since an index.
# Copyright (C) 2026 Marlon Brandao de Sousa. GPL-2. See COPYING.txt.
# ROLE: command handler for `getSpeech`.

from __future__ import annotations

from typing import TYPE_CHECKING, Any

from .... import protocol
from .command_handler import CommandHandler
from .observation import speech_entries

if TYPE_CHECKING:
	from .session_context import SessionContext


class GetSpeechHandler(CommandHandler):
	def execute(self, ctx: SessionContext, request: protocol.Request) -> Any:
		params = protocol.from_dict(protocol.GetSpeechParams, request.params)
		entries, from_index, to_index = ctx.speech_buffer.entries_since(params.sinceIndex)
		return protocol.SpeechResult(
			entries=speech_entries(entries),
			fromIndex=from_index,
			toIndex=to_index,
		)

File: bridges/voiceover/Sources/VoiceOverBridgeDomain/Controllers/Commands/GetSpeech.swift
// ROLE: controller -- `getSpeech`: everything the reader said since a bookmark.
// BUILT BY: Registry.
// READS: the session's SpeechBuffer, and nothing else.
// `fromIndex` and `toIndex` span the whole window read, while entries hold only utterances with
// words, so a caller resumes from `toIndex`.

import ScreenReaderWire

public final class GetSpeechHandler: CommandHandler {
	public init() {}

	public func execute(_ context: SessionContext, _ request: Request) throws -> any Encodable {
		let params = try request.params(as: GetSpeechParams.self)
		let read = try context.speechBuffer().entriesSince(params.sinceIndex)
		return SpeechResult(
			entries: Observation.speechEntries(read.entries),
			fromIndex: read.fromIndex,
			toIndex: read.toIndex
		)
	}
}
